File: CSharp/src/Runtime/Scripts/XEnv/argument.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using System.Threading;

namespace FrameworkEase.Universal.X.Utility
{
    public partial class XEnv
    {
        public static class Argument
        {
            internal static readonly List<KeyValuePair<string, string>> cache = new();

            internal static readonly Mutex mutex = new();

            internal static volatile bool setup;

            internal static readonly ArgumentEvaluator evaluator = new();

            public static XString.IEvaluator Evaluator => evaluator;

            internal static void EnsureSetup() { if (!setup) Setup(); }

            public static void Setup(params string[] extras)
            {
                lock (mutex)
                {
                    cache.Clear();

                    var raws = new List<string>();
                    if (extras is { Length: > 0 }) raws.AddRange(extras);
                    var cargs = Environment.GetCommandLineArgs();
                    if (cargs is { Length: > 1 }) raws.AddRange(cargs[1..]);

                    for (var i = 0; i < raws.Count; i++)
                    {
                        var arg = raws[i];
                        if (string.IsNullOrEmpty(arg) || !arg.StartsWith('-')) continue;

                        var key = arg.StartsWith("--") ? arg[2..] : arg[1..];
                        var idx = key.IndexOf('=');
                        if (idx != -1)
                        {
                            var value = key[(idx + 1)..];
                            key = key[..idx];
                            cache.Add(new KeyValuePair<string, string>(key, value));
                            continue;
                        }

                        if (i + 1 >= raws.Count || raws[i + 1].StartsWith('-'))
                        {
                            cache.Add(new KeyValuePair<string, string>(key, string.Empty));
                            continue;
                        }

                        cache.Add(new KeyValuePair<string, string>(key, raws[i + 1]));
                        i++;
                    }

                    setup = true;
                }
            }

            public static string Get(string key)
            {
                EnsureSetup();
                lock (mutex)
                {
                    foreach (var pair in cache)
                    {
                        if (pair.Key == key) return pair.Value;
                    }
                    return string.Empty;
                }
            }

            public static bool Exists(string key)
            {
                EnsureSetup();
                lock (mutex)
                {
                    foreach (var pair in cache)
                    {
                        if (pair.Key == key) return true;
                    }
                    return false;
                }
            }

            public static void Range(Func<string, string, bool> callback)
            {
                EnsureSetup();
                if (callback == null) throw new Exception("XEnv.Argument.Range: callback is null.");
                lock (mutex)
                {
                    foreach (var pair in cache)
                    {
                        if (!callback(pair.Key, pair.Value)) break;
                    }
                }
            }
        }

        internal sealed class ArgumentEvaluator : XString.IEvaluator
        {
            internal static readonly Regex Pattern = new(@"\$\{XEnv\.Argument\.([^}]+?)\}", RegexOptions.Compiled);

            public string Evaluate(string expr)
            {
                if (string.IsNullOrEmpty(expr)) return expr;
                var visited = new HashSet<string>();

                string repl(Match m)
                {
                    var keyName = m.Groups[1].Value;
                    var key = "Argument." + keyName;

                    if (keyName.Contains("${")) return $"${{Nested.{m.Value}}}";

                    if (!visited.Add(key)) return $"${{Recursive.XEnv.{key}}}";
                    try
                    {
                        var value = Argument.Get(keyName);
                        if (!string.IsNullOrEmpty(value)) return Pattern.Replace(value, repl);
                        return $"${{Unknown.XEnv.Argument.{keyName}}}";
                    }
                    finally { visited.Remove(key); }
                }

                return Pattern.Replace(expr, repl);
            }
        }
    }
}

File: Go/XEnv/argument.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XEnv

import (
	"fmt"
	"os"
	"regexp"
	"strings"
	"sync"
	"sync/atomic"

	"github.com/frameworkease/Universal.X.Utility/Go/XString"
)

type argumentScope struct {
	cache     [][2]string
	mutex     sync.RWMutex
	setup     atomic.Bool
	Evaluator XString.IEvaluator
}

type argumentEvaluator struct {
	pattern *regexp.Regexp
}

var Argument = &argumentScope{
	Evaluator: &argumentEvaluator{
		pattern: regexp.MustCompile(`\$\{XEnv\.Argument\.([^}]+?)\}`),
	},
}

func (as *argumentScope) ensureSetup() {
	if !as.setup.Load() {
		as.Setup()
	}
}

func (as *argumentScope) Setup(extras ...string) {
	as.mutex.Lock()
	defer as.mutex.Unlock()

	as.cache = nil
	raws := make([]string, 0, len(extras)+len(os.Args))
	raws = append(raws, extras...)
	raws = append(raws, os.Args[1:]...)

	for i := 0; i < len(raws); i++ {
		arg := raws[i]
		if arg == "" || !strings.HasPrefix(arg, "-") {
			continue
		}
		var key string
		if after, ok := strings.CutPrefix(arg, "--"); ok {
			key = after
		} else {
			key = strings.TrimPrefix(arg, "-")
		}
		if before, after, ok := strings.Cut(key, "="); ok {
			as.cache = append(as.cache, [2]string{before, after})
			continue
		}
		if i+1 >= len(raws) || strings.HasPrefix(raws[i+1], "-") {
			as.cache = append(as.cache, [2]string{key, ""})
			continue
		}
		as.cache = append(as.cache, [2]string{key, raws[i+1]})
		i++
	}
	as.setup.Store(true)
}

func (as *argumentScope) Get(key string) string {
	as.ensureSetup()

	as.mutex.RLock()
	defer as.mutex.RUnlock()

	for _, p := range as.cache {
		if p[0] == key {
			return p[1]
		}
	}
	return ""
}

func (as *argumentScope) Exists(key string) bool {
	as.ensureSetup()

	as.mutex.RLock()
	defer as.mutex.RUnlock()

	for _, p := range as.cache {
		if p[0] == key {
			return true
		}
	}
	return false
}

func (as *argumentScope) Range(callback func(key, value string) bool) {
	as.ensureSetup()

	if callback == nil {
		panic("XEnv.Argument.Range: callback is nil.")
	}
	as.mutex.RLock()
	defer as.mutex.RUnlock()

	for _, p := range as.cache {
		if !callback(p[0], p[1]) {
			break
		}
	}
}

func (ae *argumentEvaluator) Evaluate(expr string) string {
	if expr == "" {
		return expr
	}

	visited := make(map[string]bool)

	var repl func(string) string
	repl = func(m string) string {
		sm := ae.pattern.FindStringSubmatch(m)
		keyName := sm[1]
		key := "Argument." + keyName

		if strings.Contains(keyName, "${") {
			return fmt.Sprintf("${Nested.%v}", m)
		}

		if visited[key] {
			return fmt.Sprintf("${Recursive.XEnv.%v}", key)
		}
		visited[key] = true
		defer delete(visited, key)

		value := Argument.Get(keyName)
		if value != "" {
			return ae.pattern.ReplaceAllStringFunc(value, repl)
		}
		return fmt.Sprintf("${Unknown.XEnv.Argument.%v}", keyName)
	}

	return ae.pattern.ReplaceAllStringFunc(expr, repl)
}
